keras/src/trainers/data_adapters/data_adapter_utils.py

```diff
@@ -121,7 +121,15 @@ def class_weight_to_sample_weights(y, class_weight):
             y = np.argmax(y, axis=-1)
         else:
             y = np.squeeze(y, axis=-1)
-    y = np.round(y).astype("int32")
+    y = np.round(y)
+    if hasattr(y, "astype"):
```

Comment:
`y = np.round(y)` should have converted the tensor to a numpy tensor, no?

Reply:
I am not at the laptop right now to paste something to reproduce it, but it was failing for me. np.round was fine (but still returning a torch tensor)

Reply:
I really don't get this -- if you apply any `np` op to an input that isn't numpy (like a torch tensor) it will call `__array__` on the object to turn it into a NumPy ndarray. You should never have a torch tensor past that point

Reply:
Given this example:
```
import torch
import numpy as np
a = torch.from_numpy(np.array([1, 2.5, 3]))
b = np.round(a)
```
placing a breakpoint at the last line:

I get here:

![image](https://github.com/user-attachments/assets/8c8f3433-0a2a-44c5-a465-8cc13e970b2b)
which is [here](https://github.com/numpy/numpy/blob/3ca9e751af24426e5e70ec13a8ca6f4ea13f0758/numpy/_core/fromnumeric.py#L52)
it tries to apply the bound but:

![image](https://github.com/user-attachments/assets/a47bbbcd-44b7-4205-a17a-fe48eac392a4)

so we enter `_wrapit`
which is [here](https://github.com/numpy/numpy/blob/3ca9e751af24426e5e70ec13a8ca6f4ea13f0758/numpy/_core/fromnumeric.py#L41)

Please note that following the debugging flow into:

`conv = _array_converter(obj)
`
bring me [here](https://github.com/pytorch/pytorch/blob/9933e59c2b16c1b0475189eef63b0ff405dfd091/torch/_tensor.py#L1212), i.e.: we are in the torch domain now.
![image](https://github.com/user-attachments/assets/9f7a49a3-00a6-4f38-8e4f-d29e493ace30)

then when back to `_wrapit`
the result then is indeed a numpy array:

![image](https://github.com/user-attachments/assets/11161b85-5398-4047-991e-eb319c23c8f4)

the last call brings me to the [following](https://github.com/pytorch/pytorch/blob/9933e59c2b16c1b0475189eef63b0ff405dfd091/torch/_tensor.py#L1222):

![image](https://github.com/user-attachments/assets/c8422b27-691d-4999-ac24-a3a008e93902)
so I get back a Torch tensor again.

Reply:
```
pip list | grep torch
torch                      2.5.1+cpu
torchvision                0.20.1+cpu

[notice] A new release of pip is available: 24.0 -> 24.3.1
[notice] To update, run: pip install --upgrade pip
❯ pip list | grep numpy
numpy                      2.0.2

```

Reply:
Hi, am trying to get a better understanding and it looks like that __array_wrap__ is indeed a way to get back the original object after the numpy conversion:

https://numpy.org/doc/2.1/user/basics.interoperability.html#returning-foreign-objects

Which seems to be exactly what torch has implemented

Reply:
Ok, this is so messed up. Then we should use `ops.convert_to_numpy` at the start of the series of calls.

Reply:
Do you mean at the start of this function or even before?

Reply:
committed  'convert_to_numpy'